### PR TITLE
Fix some OpenSSL memory leaks: STACK_OF() containers freed but not their contents.

### PR DESCRIPTION
**Pull request by phad**
Also, avoid returning StatusOr<T*>(nullptr) from (private) function calls.  This causes problems and isn't required in the Cert API in any case.

### CODE REVIEW

File: cpp/log/cert.cc [hunk 1]
```diff
@@ -726,6 +730,16 @@ bool IsValidRedactedHost(const string& hostname) {
 }
 
 
+void freeOpenSslStackOfGeneralName(STACK_OF(GENERAL_NAME)* general_name_stack) {
```
Comment (AlCutter) [inline cpp/log/cert.cc]: Can these two go in an anonynamespace?

Comment (AlCutter) [inline cpp/log/cert.cc]: /free/Free/

File: cpp/log/cert.cc [hunk 2]
```diff
@@ -860,6 +874,16 @@ bool Cert::ValidateRedactionSubjectAltNameAndCN(int* dns_alt_name_count,
 }
 
 
+void freeOpenSslStackOfAsn1Integer(STACK_OF(ASN1_INTEGER)* asn1_integer_stack) {
```
Comment (AlCutter) [inline cpp/log/cert.cc]: anonyspace & /free/Free/

Comment (AlCutter) [inline cpp/log/cert.cc]: Tears of sadness that you can't do with with a templated method :/

File: cpp/log/cert.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/log/cert.cc]: Actually, maybe these could go into `util/openssl_util.{cc,h}` ?

Comment (phad) [inline cpp/log/cert.cc]: Done the anonymous namespace and the case change.

File: cpp/log/cert.cc (hunk 2 above)
Comment (phad) [inline cpp/log/cert.cc]: Done the anonymous namespace and the case change.
ACK re. tears.

File: cpp/log/cert.cc (hunk 1 above)
Comment (phad) [inline cpp/log/cert.cc]: Fixed the casing of the validateRedactionSubjectAltNames function as well, and moved that into the anonymous namespace.

File: cpp/log/cert.cc [hunk 3]
```diff
@@ -292,13 +292,14 @@ StatusOr<bool> Cert::HasBasicConstraintCATrue() const {
...
-  if (!ext_struct.ok()) {
+  if (!ext_struct.ok()
+      && ext_struct.status().CanonicalCode() != Code::NOT_FOUND) {
     // Truly odd.
     LOG(ERROR) << "Failed to check BasicConstraints extension";
     return ext_struct.status();
   }
 
-  if (!ext_struct.ValueOrDie()) {
+  if (ext_struct.status().CanonicalCode() == Code::NOT_FOUND) {
```
Comment (AlCutter) [inline cpp/log/cert.cc]: You could move this check above the one at L295, and then that one could revert to simply `if (!ext_struct.ok())`, might be a little easier to read.
Same for the other instance below at ~L322

Comment (phad) [inline cpp/log/cert.cc]: Yeah, good call - done.

**Pull request merged**